### PR TITLE
Nightly build XSIAM - search & install packs retry mechanism is broken.

### PR DESCRIPTION
**Pull request by kobymeir**
## Status
- [] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-6947
nightly: https://code.pan.run/xsoar/content/-/pipelines/5514298

## Description
Nightly build XSIAM - search & install packs retry mechanism is broken.
fixing pathlib ruff issues.

### CODE REVIEW

File: Tests/Marketplace/search_and_install_packs.py [hunk 1]
```diff
@@ -303,10 +311,19 @@ def install_packs_private(client: demisto_client,
                                  test_pack_path=test_pack_path)
 
 
+def get_error_ids(body: str) -> set[str] | None:
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:314]: No need to annotate is as a static method?

File: Tests/Marketplace/search_and_install_packs.py [hunk 2]
```diff
@@ -303,10 +311,19 @@ def install_packs_private(client: demisto_client,
...
+    with contextlib.suppress(JSONDecodeError):
+        response_info = json.loads(body)
+        return {error["id"] for error in response_info.get("errors", [])}
+    return None
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:318]: Why not return an empty list? (it takes more memory, but the return value will be `set[str]`

File: Tests/Marketplace/search_and_install_packs.py [hunk 3]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
         host (str): The server URL.
         packs_to_install (list): A list of the packs to install.
         request_timeout (int): Timeout settings for the installation request.
+        attempts_count (int): The number of attempts to install the packs.
+        sleep_interval (int): The sleep interval between install attempts.
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:340]:
```suggestion
        sleep_interval (int): The sleep interval, in seconds, between install attempts.
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 4]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
         host (str): The server URL.
         packs_to_install (list): A list of the packs to install.
         request_timeout (int): Timeout settings for the installation request.
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:338]:
```suggestion
        request_timeout (int): Timeout setting, in seconds, for the installation request.
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 5]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
-                                                                                    body={'packs': packs,
-                                                                                          'ignoreWarnings': True},
-                                                                                    accept='application/json',
-                                                                                    _request_timeout=request_timeout,
-                                                                                    response_type='object')
-
-                if status_code in range(200, 300) and status_code != 204:
-                    packs_data = [{'ID': pack.get('id'), 'CurrentVersion': pack.get('currentVersion')} for pack in response_data]
+                logging.info(f"Installing packs {', '.join([p.get('id') for p in packs_to_install])} on server {host}. "
+                             f"Attempt:{attempts_count - attempt}/{attempts_count}")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:347]:
```suggestion
                             f"Attempt: {attempts_count - attempt}/{attempts_count}")
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 6]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+                                                                                     response_type='object')
+
+                if 200 <= status_code < 300 and status_code != 204:
+                    packs_data = [{'ID': pack.get('id'), 'CurrentVersion': pack.get('currentVersion')} for pack in response]
                     logging.success(f'Packs were successfully installed on server {host}')
                     logging.debug(f'The packs that were successfully installed on server {host}:\n{packs_data}')
                     break
 
+                if not attempt:
+                    raise Exception(f"Got bad status code: {status_code}, headers:{headers}")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:364]: Can `headers` contain sensitive info?

Also:
```suggestion
                    raise Exception(f"Got bad status code: {status_code}, headers: {headers}")
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 7]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+                    if 'timeout awaiting response' in ex.body:
+                        if '/packs/' in ex.body:
+                            pack_id = get_pack_id_from_error_with_gcp_path(ex.body)
+                            raise Exception(f"timeout awaiting response headers while trying to install pack {pack_id}") from ex
+
+                        raise Exception("timeout awaiting response headers while trying to install, "
+                                        "couldn't determine pack id.") from ex
+
+                    if 'Item not found' in ex.body:
+                        raise Exception(f'Item not found error, headers:{ex.headers}.') from ex
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:393]:
```suggestion
                        raise Exception(f'Item not found error while trying to install packs {packs_to_install}, headers:{ex.headers}.') from ex
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 8]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+
+                    # Unknown exception reason, re-raise.
+                    raise Exception(f"Got {ex.status} from server, message:{ex.body}, headers:{ex.headers}") from ex
+            except (HTTPError, HTTPWarning) as http_ex:
+                if not attempt:
+                    raise Exception("Failed to perform http request to the server") from http_ex
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:399]:
```suggestion
                    raise Exception("Failed to perform http request to the server to install packs {packs_to_install}") from http_ex
```

File: Tests/Marketplace/search_and_install_packs.py [hunk 9]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+
+            # There are more attempts available, sleep and retry.
+            logging.debug(f"failed to install pack, sleeping for {sleep_interval} seconds.")
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:402]: Add the pack id (or packs id's)

File: Tests/Marketplace/search_and_install_packs.py [hunk 10]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+            sleep(sleep_interval)
     except Exception as e:
         logging.exception(f'The request to install packs has failed. Additional info: {str(e)}')
```
Comment (tkatzir) [inline Tests/Marketplace/search_and_install_packs.py:405]:
```suggestion
        logging.exception(f'The request to install packs {packs_to_install} has failed. Additional info: {str(e)}')
```

**Review by tkatzir — COMMENTED** [review]
Very good! :-)
Some notes and comments.

File: Tests/Marketplace/search_and_install_packs.py [hunk 11]
```diff
@@ -228,21 +235,22 @@ def find_malformed_pack_id(body: str) -> List:
...
             else:
-                malformed_pack_pattern = re.compile(r'invalid version [0-9.]+ for pack with ID ([\w_-]+)')
-                malformed_pack_id = malformed_pack_pattern.findall(str(error))
-                if malformed_pack_id and error:
-                    malformed_ids.extend(malformed_pack_id)
+                # the error is returned as a list of error
+                errors_info = response_info.get('errors', [])
+            for error in errors_info:
+                if 'pack id: ' in error:
+                    malformed_ids.extend(error.split('pack id: ')[1].replace(']', '').replace('[', '').replace(
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:247]: you're within that method, will add UT for it.

File: Tests/Marketplace/search_and_install_packs.py [hunk 12]
```diff
@@ -228,21 +235,22 @@ def find_malformed_pack_id(body: str) -> List:
...
+        with contextlib.suppress(JSONDecodeError):
+            response_info = json.loads(body)
+            if error_info := response_info.get('error'):
+                errors_info = [error_info]
             else:
-                malformed_pack_pattern = re.compile(r'invalid version [0-9.]+ for pack with ID ([\w_-]+)')
-                malformed_pack_id = malformed_pack_pattern.findall(str(error))
-                if malformed_pack_id and error:
-                    malformed_ids.extend(malformed_pack_id)
+                # the error is returned as a list of error
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:243]: not my typo :)

File: Tests/Marketplace/search_and_install_packs.py [hunk 13]
```diff
@@ -228,21 +235,22 @@ def find_malformed_pack_id(body: str) -> List:
...
+                errors_info = response_info.get('errors', [])
+            for error in errors_info:
+                if 'pack id: ' in error:
+                    malformed_ids.extend(error.split('pack id: ')[1].replace(']', '').replace('[', '').replace(
+                        ' ', '').split(','))
+                else:
+                    malformed_pack_pattern = re.compile(r'invalid version [0-9.]+ for pack with ID ([\w_-]+)')
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:250]: sure, will optimize.

File: Tests/Marketplace/search_and_install_packs.py (hunk 1 above)
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:314]: it's not within a class

File: Tests/Marketplace/search_and_install_packs.py (hunk 2 above)
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:318]: sure

File: Tests/Marketplace/search_and_install_packs.py [hunk 14]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+                        raise Exception(f"malformed packs:{malformed_ids}") from ex
 
+                    # We've more attempts, retrying without tho malformed packs.
+                    SUCCESS_FLAG = False
+                    logging.error(f"Unable to install malformed packs:{malformed_ids}, retrying without them.")
+                    packs_to_install = [pack for pack in packs_to_install if pack['id'] not in malformed_ids]
+
+                if (error_ids := get_error_ids(ex.body)) and TEST_MODELING_RULE_ERROR_CODE in error_ids:
+                    # If we got this error code, it means that the modeling rules are not valid, exiting install flow.
+                    raise Exception("Test modeling rules for pack are not valid") from ex
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:381]: not that i can find, it need to simulate it to get the exact error, this is not trivial to produce.

File: Tests/Marketplace/search_and_install_packs.py [hunk 15]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+
+                if not attempt:  # exhausted all attempts, understand what happened and exit.
+                    if 'timeout awaiting response' in ex.body:
+                        if '/packs/' in ex.body:
+                            pack_id = get_pack_id_from_error_with_gcp_path(ex.body)
+                            raise Exception(f"timeout awaiting response headers while trying to install pack {pack_id}") from ex
+
+                        raise Exception("timeout awaiting response headers while trying to install, "
+                                        "couldn't determine pack id.") from ex
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:390]: ignoring, the pack ids are printed in the outer exception block

File: Tests/Marketplace/search_and_install_packs.py (hunk 8 above)
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:399]: ignoring, the pack ids are printed in the outer exception block

File: Tests/Marketplace/search_and_install_packs.py [hunk 16]
```diff
@@ -319,86 +336,71 @@ def install_packs(client: demisto_client,
...
+
+                    if 'Item not found' in ex.body:
+                        raise Exception(f'Item not found error, headers:{ex.headers}.') from ex
+
+                    # Unknown exception reason, re-raise.
+                    raise Exception(f"Got {ex.status} from server, message:{ex.body}, headers:{ex.headers}") from ex
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:396]: ignoring, the pack ids are printed in the outer exception block

File: Tests/Marketplace/search_and_install_packs.py (hunk 9 above)
Comment (kobymeir) [inline Tests/Marketplace/search_and_install_packs.py:402]:
```suggestion
            logging.debug(f"failed to install packs: {packs_to_install}, sleeping for {sleep_interval} seconds.")
```

Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:402]: added

File: Tests/Marketplace/search_and_install_packs.py (hunk 7 above)
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:393]: ignoring, the pack ids are printed in the outer exception block

File: Tests/Marketplace/search_and_install_packs.py [hunk 17]
```diff
@@ -157,13 +162,15 @@ def get_pack_dependencies(client: demisto_client, pack_data: dict, lock: Lock):
             return []
         msg = response_data.get('message', '')
         raise Exception(f'status code {status_code}\n{msg}\n')
+    except ApiException as api_ex:
+        with lock:
+            SUCCESS_FLAG = False
+        logging.exception(f"The request to get pack {pack_id} dependencies has failed, Got {api_ex.status} from server, "
+                          f"message:{api_ex.body}, headers:{api_ex.headers}")
```
[the remark this answers is not in the document]
Comment (kobymeir) [reply Tests/Marketplace/search_and_install_packs.py:169]: @tkatzir  potentially yes, Anar wanted it printed, any suggestion here?

File: Tests/Marketplace/search_and_install_packs.py [hunk 18]
```diff
@@ -318,89 +335,74 @@ def install_packs(client: demisto_client,
...
+                        raise Exception(f"malformed packs: {malformed_ids}") from ex
 
+                    # We've more attempts, retrying without tho malformed packs.
+                    SUCCESS_FLAG = False
+                    logging.error(f"Unable to install malformed packs: {malformed_ids}, retrying without them.")
+                    packs_to_install = [pack for pack in packs_to_install if pack['id'] not in malformed_ids]
+
+                if (error_ids := get_error_ids(ex.body)) and TEST_MODELING_RULE_ERROR_CODE in error_ids:
+                    # If we got this error code, it means that the modeling rules are not valid, exiting install flow.
+                    raise Exception(f"Test modeling rules for packs {packs_to_install} are not valid") from ex
```
Comment (daryakoval) [inline Tests/Marketplace/search_and_install_packs.py:381]: This is not the right error description. 
The error code is when WLM Task ID fails. WLM Task is running validations on modeling rules and datasets.
Leave the error code in the log. We are using this error code to find the logs: https://confluence-dc.paloaltonetworks.com/display/DemistoContent/Logs+in+the+GCP#LogsintheGCP-AdvancedDebugging (Please see the video). Without the error code the developer should go to the code and find the error code and only then go to GCP. 

```suggestion
                    raise Exception(f"Failed to install: {packs_to_install}. Got [{WLM_TASK_FAILED_ERROR_CODE}] error code - Modeling rules and Dataset validations failed. Please look at GCP logs to understand why it failed.") from ex
                    
```

**Comment by kobymeir** [comment]
Forced merged reasons:
- MicrosoftCore test modeling rule is broken.
- Error in API call [404] - Not Found - also happens in nightly.
- TestUploadFlow - flakey tests

**Review by tkatzir — APPROVED** [review]
Great!

**Pull request merged**